src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -156,59 +145,99 @@ internal static TypeSymbolWithAnnotations Create(INonNullTypesContext nonNullTyp
             {
                 return default;
             }
-            if (!isAnnotated && typeSymbol.IsNullableType())
+
+            return Create(typeSymbol, nullableAnnotation: isAnnotated ? NullableAnnotation.Nullable : nonNullTypesContext.NonNullTypes == true ? NullableAnnotation.NotNullable : NullableAnnotation.Unknown,
+                          customModifiers.NullToEmpty());
+        }
+
+        internal static TypeSymbolWithAnnotations Create(TypeSymbol typeSymbol, NullableAnnotation nullableAnnotation = NullableAnnotation.Unknown, ImmutableArray<CustomModifier> customModifiers = default)
+        {
+            if (typeSymbol is null)
+            {
+                return default;
+            }
+
+            if (!nullableAnnotation.IsAnyNullable() && typeSymbol.IsNullableType())
+            {
+                // int?, T? where T : struct (add annotation)
+                nullableAnnotation = NullableAnnotation.Nullable;
+            }
+
+            return CreateNonLazyType(typeSymbol, nullableAnnotation, customModifiers.NullToEmpty());
+        }
+
+        internal bool IsPossiblyNullableReferenceTypeTypeParameter()
+        {
+            return NullableAnnotation == NullableAnnotation.NotNullable && TypeSymbol.IsPossiblyNullableReferenceTypeTypeParameter();
+        }
+
+        internal NullableAnnotation GetValueNullableAnnotation()
```

Comment:
>GetValueNullableAnnotation [](start = 36, length = 26)

nit: I found this method name confusing. What does "Value" refer to?
Perhaps `GetNullableAnnotationForAnalysis` would be clearer?

Reply:
> I found this method name confusing. What does "Value" refer to?

It refers to the fact that we are talking about nullability of a value if this type. There is a distinction between a value and a target of an assignment. Both are involved into analysis, but handled differently even if they have the same type.

---
In reply to: [231236701](https://github.com/dotnet/roslyn/pull/30913#discussion_r231236701) [](ancestors = 231236701)